src/Streamly/Coreutils/Uniq.hs

```diff
@@ -0,0 +1,149 @@
+             else go (i + 1) x
+    isChar = not . isSpace
+
+
+-- slice offset length
+slice :: Int -> Int -> String -> String
+slice off len = take len . drop off
+
+
+compareIgnCase :: String -> String -> Bool
+compareIgnCase str1 str2 | length str1 /= length str2 = False
+compareIgnCase (x1:xs) (y1:ys) =
+    (toLower x1 == toLower y1) && compareIgnCase xs ys
+compareIgnCase [] [] = True
+
+
+compareWith :: (String -> String -> Bool) -> String -> String -> Bool
+compareWith eq s1 s2 = eq s1 s2
+
+
+-- Skip X terms
+-- Skip Y chars
+-- Take Z chars
+compareXYZ :: Bool -> Int -> Int -> Int -> String -> String -> Bool
+compareXYZ ignCase x y z str1 str2 =
+      compareWith compareFunc (slice (ix1 + y) z str1) (slice (ix2 + y) z str2)
+  where
+    ix1 = indexAfterSkippingNWords x str1
+    ix2 = indexAfterSkippingNWords x str2
+    compareFunc =
+        if ignCase
+        then compareIgnCase
+        else (==)
+
+
+
+compareUsingOptions :: UniqOptions -> (String -> String -> Bool)
+compareUsingOptions opt = compareXYZ (ignoreCase opt) (skipFields opt) (skipChar opt) (checkChar opt)
+
+
+-- Skip X terms
+-- Skip Y chars
+-- Take Z chars
+getRepetition :: (IsStream t, Monad m) => (String -> String -> Bool) -> t m String -> t m (Int, String)
+getRepetition comparator = S.groupsBy comparator
+                              (FL.mkPureId (\(i, a) s -> if a == ""
+                                                         then (i + 1, s)
+                                                         else (i + 1, a)) (0, ""))
+
+
+onlyDuplicate :: (IsStream t, Monad m) => t m (Int, String) -> t m String
+onlyDuplicate = S.concatMap (\(i, x) -> if i > 1
+                                        then S.replicate i x
+                                        else S.nil)
+
+
+onlyRepeated :: (IsStream t, Monad m) => t m (Int, String) -> t m String
+onlyRepeated = S.concatMap (\(i, x) -> if i > 1
+                                       then S.replicate 1 x
+                                       else S.nil)
+
+
+onlyUnique :: (IsStream t, Monad m) => t m (Int, String) -> t m String
+onlyUnique = S.concatMap (\(i, x) -> if i == 1
+                                     then S.replicate i x
+                                     else S.nil)
+
+
+uniqStrm :: (IsStream t, Monad m, MonadIO m) => (String -> String -> Bool) -> t m String -> t m String
+uniqStrm comparator strm = onlyUnique
+   $ getRepetition comparator strm
+
+
+duplicateStrm :: (IsStream t, Monad m, MonadIO m) => (String -> String -> Bool) -> t m String -> t m String
+duplicateStrm comparator strm = onlyDuplicate
+   $ getRepetition comparator strm
+
+
+repeatedStrm :: (IsStream t, Monad m, MonadIO m) => (String -> String -> Bool) -> t m String -> t m String
+repeatedStrm comparator strm = onlyRepeated
+   $ getRepetition comparator strm
+
+
+displayCounts :: (IsStream t, Monad m, MonadIO m) => (String -> String -> Bool) -> ((Int, String) -> Bool) -> t m String -> t m (Int, String)
```

Comment:
What is this used for?

Reply:
When we want to provide counts to the user, at that time, the returned stream should have type `t m (Int, String)`

Reply:
& the predicate `((Int, String) -> Bool)` is to filter which kind of lines (unique, repeated, distinct) is needed in the output stream